pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -762,16 +768,154 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+	}
+
+	mark := linux_defaults.MagicMarkEgress
 	mark |= int(uint32(secId&0xFFFF)<<16 | uint32((secId&0xFF0000)>>16))
 	err := unix.SetsockoptInt(fd, unix.SOL_SOCKET, unix.SO_MARK, mark)
 	if err != nil {
 		return fmt.Errorf("error setting SO_MARK: %w", err)
 	}
+
+	if err := unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEADDR, 1); err != nil {
+		return fmt.Errorf("setsockopt(SO_REUSEADDR) failed: %w", err)
+	}
+	if !option.Config.EnableBPFTProxy {
+		if err := unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1); err != nil {
+			return fmt.Errorf("setsockopt(SO_REUSEPORT) failed: %w", err)
+		}
+	}
+
 	return nil
 }
 
+var errNoReader = errors.New("Reader stopped")
+
+type Response struct {
+	*dns.Msg
+	err error
+}
+
+// A Client keeps state for concurrent transactions on the same upstream connection.
+type Client struct {
+	serverAddr string
+
+	dns.Client
+	refcount int // protected by proxy's lock
+
+	// this mutex protects writes on 'conn' and all access to 'reqs'
+	lock.RWMutex
+	reqs map[uint16]chan Response // outstanding requests
+
+	// 'readerLock' mutex is used to serialize reads on 'conn'. It is always taken and released
+	// while holding the main lock but the main lock can be released and re-acquired while
+	// holding 'readerLock' mutex.
+	readerLock lock.Mutex
+
+	// Client's connection shared among all requests from the same source address/port. The
+	// locks above are used to serialize reads and writes on this connection, but reads and
+	// writes can happen at the same time.
+	conn *dns.Conn
+}
+
+// ExchangeAsync writes the request to the Client's connection and co-operatively
+// reads responses from the connection and distributes them to the requestors.
+// At most one caller is reading from Client's connection at any time.
+func (c *Client) ExchangeAsync(m *dns.Msg) (r *dns.Msg, rtt time.Duration, err error) {
+	// Lock allows only one request to be written at a time, but that can happen
+	// concurrently with reading.
+	c.Lock()
+	defer c.Unlock()
+	if _, exists := c.reqs[m.Id]; exists {
+		return nil, 0, fmt.Errorf("duplicate request: %d", m.Id)
+	}
+
+	// Dial if needed
+	if c.conn == nil {
+		c.conn, err = c.Dial(c.serverAddr)
+		if err != nil {
+			return nil, 0, fmt.Errorf("failed to dial connection to %v: %w", c.serverAddr, err)
+		}
+	}
+
+	// Send while holding the client lock, as dns.Client is not made to be usable from
+	// concurrent goroutines.
+	start := time.Now()
+	err = c.SendContext(context.Background(), m, c.conn, start)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Create channel for the response with buffer of one, so that write to it
+	// does not block if we happen to do it ourselves.
+	ch := make(chan Response, 1)
+	c.reqs[m.Id] = ch
+
+	// Wait for the response
+	var resp Response
+	for {
+		// Try taking the reader lock
+		if c.readerLock.TryLock() {
+			// We are responsible for reading responses for all users
```

Comment:
Tried it, it is structurally more complex, and adds spawning a new goroutine for each proxied DNS request, something that you have tried to weed off on the agent regeneration side of things.

Reply:
I said goroutine per client above, of course per request would be excessive.

Reply:
Clean and nice pattern, thank you.

Due to the restrictions placed on the proxy by in needing to be transparent, i.e., use the original source address and port in the DNS client connection, this pattern leads to, on average, a goroutine per request for dual stack pods, and 2 goroutines per request for single stack  pods. Why is this? Well,

- Since we must use the original source IP address, and we must also intercept the responses, we instruct the networking namespace of the dnsproxy (host) to receive packets destined to a non-local (foreign) IP address via the IP_TRANSPARENT socket option.
- in order to not hijack some random by-standing traffic going to the original pod, we must also use the original port number
- (DNS) clients use ephemeral source ports, i.e., the port can be different in every request. Typically, a DNS resolver library uses the same ephemeral port only to requests from a single "gethostbyname" API call, or equivalent.
- To be able to receive responses to the ephemeral source port, we must have a socket bound to that address:port (for UDP), or a connection from that address:port to the DNS server address:port (for TCP).
- This leads to a new DNS client and socket for every different original source address - ephemeral port pair we see. We also need to make sure these were actually used to communicate with the DNS server, so we use the whole 5-tuple as a key

Why can't we keep DNS clients pooled and ready to receive traffic between client requests then?
- We have no guarantees that the source pod will keep on using the same ephemeral port in future. We've had upstream socket bind errors (in Envoy, where we have operated in this mode for years already) when a client pod has rapidly cycled through its ephemeral port space, e.g. when performing netperf CRR or similar performance tests.
- We could try keep the client and its bound socket around for some minimal time to save resources when a DNS resolver is enumerating through its domain suffix list, where is seems likely that the same source ephemeral port is going to be reused until the resolver gets an actual result with an IPv4/6 address or quits trying. This would be something we currently don't do and is prone to socket bind errors, so I'd like to leave this for later.
- So the client socket can not be left lingering around, as it causes network traffic destined for the source pod to be intercepted to the dnsproxy, whcih is exactly what we want until a DNS response has been received.

Reply:
Thanks for the detailed explanation. This is so great I wonder if we should add some of this to code comments?
Given the ephemeral nature and the amount of polish and review, I'm fine with keeping it this way.

Reply:
+1 to add this a comment. Thanks for clarifying this!